### PR TITLE
add vmsyscall crate

### PR DESCRIPTION
**Pull request by haraldh**

### CODE REVIEW

File: vmsyscall/src/error.rs [hunk 1]
```diff
@@ -0,0 +1,321 @@
+//! errno definition
+//!
+//! copied from Redox
```
[the remark this answers is not in the document]
Comment (haraldh) [reply vmsyscall/src/error.rs]: yeah, we might

[the remark this answers is not in the document]
Comment (npmccallum) [reply vmsyscall/src/error.rs]: @haraldh I agree with @steveeJ . No code copying please. Should we reject this PR altogether?

[the remark this answers is not in the document]
Comment (haraldh) [reply vmsyscall/src/error.rs]: I'll update the PR

[the remark this answers is not in the document]
Comment (haraldh) [reply vmsyscall/src/error.rs]: To clarify, this is only one file out of the whole crate
https://gitlab.redox-os.org/redox-os/syscall
and I just need the errno consts

[the remark this answers is not in the document]
Comment (haraldh) [reply vmsyscall/src/error.rs]: maybe I should just use https://crates.io/crates/errno-codes and add more dependencies

**Comment by haraldh** [comment]
here we go again

File: vmsyscall/src/error.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (haraldh) [reply vmsyscall/src/error.rs]: meh, `errno-codes` is not no_std

File: vmsyscall/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+    Mmap(Result<usize, Error>),
+    Mremap(Result<usize, Error>),
+    Munmap(Result<i32, Error>),
+    Mprotect(Result<i32, Error>),
+}
+
+#[derive(Debug, Serialize, Deserialize, PartialEq)]
+pub enum Error {
+    Errno(i32),
+    NotImplemented,
```
Comment (npmccallum) [inline vmsyscall/src/lib.rs]: Isn't there already `ENOSYS`?

Comment (haraldh) [reply vmsyscall/src/lib.rs]: right

**Comment by haraldh** [comment]
> Why is this dual licensed?

all my work was dual licensed... can change

**Comment by haraldh** [comment]
updated licencse and added errno and syscall consts

**Comment by npmccallum** [comment]
@haraldh I need the syscall and errno consts too. Can we make a separate crate for these? Should we make them enums?

**Comment by haraldh** [comment]
removed the syscall and errno consts

**Comment by haraldh** [comment]
includes no main Cargo.toml and Cargo.lock changes

**Comment by haraldh** [comment]
added Cargo.toml ... otherwise travis is useless

File: vmsyscall/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! syscall serialize/deserialize
+//!
+//! Currently it uses a hard coded page and an I/O trigger.
+//! We might want to switch to MMIO.
+
+#![no_std]
+
```
Comment (steveeJ) [inline vmsyscall/src/lib.rs]: Maybe ` #![deny(missing_docs)]` and add docs to satisfy it?

Comment (haraldh) [reply vmsyscall/src/lib.rs]: done

File: vmsyscall/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,71 @@
+// Copyright 2019 Red Hat, Inc.
```
[the remark this answers is not in the document]
Comment (haraldh) [reply vmsyscall/src/lib.rs]: done

**Comment by npmccallum** [comment]
@haraldh I think this needs a rebase.

File: vmsyscall/Cargo.toml [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+authors = ["Harald Hoyer <harald@redhat.com>"]
+license = "Apache-2.0"
+edition = "2018"
+
+categories = [
+    "no-std",
+]
+
+keywords = [
+    "errno",
```
Comment (npmccallum) [inline vmsyscall/Cargo.toml]: `errno` is now a different crate.

File: vmsyscall/Cargo.toml [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+    "syscall",
+    "no_std",
+]
+
+[dependencies]
+serde = { version = "1.0", default-features = false, features = ["derive"] }
```
Comment (npmccallum) [inline vmsyscall/Cargo.toml]: I don't remember how cargo handles `no_std` in this case. But I think the consumer of this crate will also have to specify `no_std` as a feature for `serde`. This is just a note for my future recall. :)

File: vmsyscall/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! syscall serialize/deserialize
+//!
+//! Currently it uses a hard coded page and an I/O trigger.
+//! We might want to switch to MMIO.
+
+#![deny(missing_docs)]
```
Comment (npmccallum) [inline vmsyscall/src/lib.rs]:
```suggestion
#![deny(missing_docs)]
#![deny(clippy:all)]
```

File: vmsyscall/Cargo.toml (hunk 6 above)
Comment (haraldh) [reply vmsyscall/Cargo.toml]: At least for serde_cbor it works this way. No_std only has to be defined by the implementing serialize/deserialize crate...

File: vmsyscall/Cargo.toml (hunk 5 above)
Comment (haraldh) [reply vmsyscall/Cargo.toml]: Right

File: vmsyscall/Cargo.toml (hunk 6 above)
Comment (haraldh) [reply vmsyscall/Cargo.toml]: Default is "std"
```
[features]
default = ["std"]

# Provide derive(Serialize, Deserialize) macros.
derive = ["serde_derive"]

# Provide impls for common standard library types like Vec<T> and HashMap<K, V>.
# Requires a dependency on the Rust standard library.
std = []
```

File: vmsyscall/src/lib.rs (hunk 7 above)
Comment (haraldh) [reply vmsyscall/src/lib.rs]: done

File: vmsyscall/Cargo.toml (hunk 5 above)
Comment (haraldh) [reply vmsyscall/Cargo.toml]: removed

File: vmsyscall/src/lib.rs [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+    /// int madvise(void *addr, size_t length, int advice);
+    Madvise {
+        /// see madvise(2)
+        addr: usize,
+        /// see madvise(2)
+        length: usize,
+        /// see madvise(2)
+        advice: i32,
+    },
+    /// void *mmap(void *addr, size_t length, int prot, int flags, …);
```
Comment (steveeJ) [inline vmsyscall/src/lib.rs]: What does `…` hint at?

File: vmsyscall/src/lib.rs [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+    /// int mprotect(void *addr, size_t len, int prot);
+    Mprotect {
+        /// see mprotect(2)
+        addr: usize,
+        /// see mprotect(2)
+        length: usize,
+        /// see mprotect(2)
+        prot: i32,
+    },
+    // Todo: extend with needed hypervisor proxy syscalls
```
Comment (steveeJ) [inline vmsyscall/src/lib.rs]: Who do we find out which syscalls are needed? Are these the ones required by the runtime which is added in #139?

File: vmsyscall/src/lib.rs (hunk 8 above)
Comment (haraldh) [reply vmsyscall/src/lib.rs]: Left out unneeded parameters

File: vmsyscall/src/lib.rs (hunk 9 above)
Comment (haraldh) [reply vmsyscall/src/lib.rs]: For the sev kernel case we only proxy those, which the kernel can't handle itsself.. like network related syscalls like socket(2). And if the kernel does not want to do extensive memory handling, we want to proxy the mmap family, too.

Comment (haraldh) [reply vmsyscall/src/lib.rs]: currently this is just a skeleton crate, which will be extended and stripped a lot in the future.

File: vmsyscall/src/lib.rs [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+//!
+//! Currently it uses a hard coded page and an I/O trigger.
+//! We might want to switch to MMIO.
+
+#![deny(missing_docs)]
+#![deny(clippy::all)]
+#![no_std]
+
+/// The IO trigger port for vm syscall proxy
+pub const TRIGGER_PORT: u16 = 0xFF;
```
Comment (npmccallum) [inline vmsyscall/src/lib.rs]: Sorry for one more question. `TRIGGER_PORT` seems to me a VM specific detail which we won't use in process-based TEEs. Is this crate the right place for it?

Comment (haraldh) [reply vmsyscall/src/lib.rs]: oh .. right..

Comment (haraldh) [reply vmsyscall/src/lib.rs]: removed

**Pull request merged**